src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+            (region.imageExtent.width + region.imageOffset.x <= subresourceSize.x) &&
+            (region.imageExtent.height + region.imageOffset.y <= subresourceSize.y) &&
+            (region.imageExtent.depth + region.imageOffset.z <= subresourceSize.z);
+        assert(isImageExtentAndOffsetValid);
+
+        auto imageExtent = core::vector3du32_SIMD(region.imageExtent.width, region.imageExtent.height, region.imageExtent.depth);
+        auto imageExtentInBlocks = dstImageTexelBlockInfo.convertTexelsToBlocks(imageExtent);
+        auto imageExtentBlockStridesInBytes = dstImageTexelBlockInfo.convert3DBlockStridesTo1DByteStrides(imageExtentInBlocks);
+
+        if(i == currentRegion)
+        {
+            auto remainingBlocksInRow = imageExtentInBlocks.x - currentBlockInRow;
+            auto remainingRowsInSlice = imageExtentInBlocks.y - currentRowInSlice;
+            auto remainingSlicesInLayer = imageExtentInBlocks.z - currentSliceInLayer;
+            auto remainingLayersInRegion = region.imageSubresource.layerCount - currentLayerInRegion;
+
+            if(currentBlockInRow == 0 && currentRowInSlice == 0 && currentSliceInLayer == 0 && remainingLayersInRegion > 0)
+                memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * remainingLayersInRegion;
+            else if (currentBlockInRow == 0 && currentRowInSlice == 0 && currentSliceInLayer > 0)
+            {
+                memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[2] * remainingSlicesInLayer;
+                if(remainingLayersInRegion > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * (remainingLayersInRegion - 1u);
+            }
+            else if (currentBlockInRow == 0 && currentRowInSlice > 0)
+            {
+                memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[1] * remainingRowsInSlice;
+                if(remainingSlicesInLayer > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[2] * (remainingSlicesInLayer - 1u);
+                if(remainingLayersInRegion > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * (remainingLayersInRegion - 1u);
+            }
+            else if (currentBlockInRow > 0)
+            {
+                memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[0] * remainingBlocksInRow;
+                if(remainingRowsInSlice > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[1] * (remainingRowsInSlice - 1u);
+                if(remainingSlicesInLayer > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[2] * (remainingSlicesInLayer - 1u);
+                if(remainingLayersInRegion > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * (remainingLayersInRegion - 1u);
+            }
```

Comment:
you need an alignUp before every `+=`, see https://github.com/Devsh-Graphics-Programming/Nabla/pull/389/files#r955045146

Reply:
You're right